File: packages/go/internal/seguranca/ataques/ataque_diferencial_keystream.go
package ataques

import (
	"encoding/hex"
	"fmt"

	"criptografia/internal/seguranca"
)

// OpcoesDiferencialKeystream configura o AtaqueDiferencialKeystream.
type OpcoesDiferencialKeystream struct {
	Amostras     int
	TamanhoBloco int
}

// AtaqueDiferencialKeystream faz criptanálise diferencial no keystream com uma
// diferença FIXA (1 bit) no IV.
type AtaqueDiferencialKeystream struct {
	amostras     int
	tamanhoBloco int
}

// NovoAtaqueDiferencialKeystream cria o ataque. Sem argumento, usa
// amostras = 2000 e tamanhoBloco = 32.
func NovoAtaqueDiferencialKeystream(opcoes ...OpcoesDiferencialKeystream) *AtaqueDiferencialKeystream {
	o := OpcoesDiferencialKeystream{Amostras: 2000, TamanhoBloco: 32}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueDiferencialKeystream{amostras: o.Amostras, tamanhoBloco: o.TamanhoBloco}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueDiferencialKeystream) Nome() string {
	return "Diferencial do keystream (delta fixo no IV)"
}

// Executar roda o ataque.
func (a *AtaqueDiferencialKeystream) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	chave := seguranca.RandomBytes(len(alvo.ChaveDeTeste()))
	tamanhoIv := alvo.TamanhoIv()

	delta := make([]byte, tamanhoIv)
	delta[seguranca.RandomInt(0, tamanhoIv-1)] = byte(1 << seguranca.RandomInt(0, 7))

	totalBits := a.tamanhoBloco * 8
	sempreZero := make([]bool, totalBits)
	sempreUm := make([]bool, totalBits)
	for i := range sempreZero {
		sempreZero[i] = true
		sempreUm[i] = true
	}
	deltas := map[string]bool{}

	for i := 0; i < a.amostras; i++ {
		iv := seguranca.RandomBytes(tamanhoIv)
		iv2 := a.xorBytes(iv, delta)

		ks1 := alvo.GerarKeystreamBruto(chave, iv, "enc", a.tamanhoBloco)
		ks2 := alvo.GerarKeystreamBruto(chave, iv2, "enc", a.tamanhoBloco)
		d := a.xorBytes(ks1, ks2)
		deltas[hex.EncodeToString(d)] = true

		for p := 0; p < len(d); p++ {
			by := d[p]
			for b := 0; b < 8; b++ {
				idx := p*8 + b
				if ((by >> b) & 1) == 1 {
					sempreZero[idx] = false
				} else {
					sempreUm[idx] = false
				}
			}
		}
	}

	bitsFixos := 0
	for idx := 0; idx < totalBits; idx++ {
		if sempreZero[idx] || sempreUm[idx] {
			bitsFixos++
		}
	}
	colisoes := a.amostras - len(deltas)

	vulneravel := bitsFixos > 0 || colisoes > 0
	severidade := seguranca.SevInfo
	if vulneravel {
		severidade = seguranca.SevAlta
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: vulneravel,
		Severidade: severidade,
		Detalhes: fmt.Sprintf(
			"delta de 1 bit no IV: %d bit(s) de saída fixo(s), %d diferencial(is) repetido(s) em %d amostras",
			bitsFixos, colisoes, a.amostras,
		),
		Dados: map[string]any{"bits_fixos": bitsFixos, "colisoes": colisoes},
	}, nil
}

// xorBytes faz XOR byte a byte de dois buffers.
func (a *AtaqueDiferencialKeystream) xorBytes(a2, b []byte) []byte {
	out := make([]byte, len(a2))
	for i := 0; i < len(a2); i++ {
		out[i] = a2[i] ^ b[i]
	}
	return out
}

File: packages/julia/src/Seguranca/Ataques/AtaqueDiferencialKeystream.jl
"""
Criptanálise diferencial no keystream: com uma diferença FIXA (1 bit) no
IV, coleta as diferenças de saída entre ks(iv) e ks(iv ^ delta). Numa cifra
boa, essas diferenças são uniformes e únicas. Sinaliza:
 - bits de saída que NUNCA mudam (ou mudam sempre) para essa diferença;
 - diferenças de saída que se REPETEM (diferencial de alta probabilidade),
   que é o insumo básico de um ataque diferencial.
"""
struct AtaqueDiferencialKeystream <: Ataque
    amostras::Int
    tamanho_bloco::Int
end

AtaqueDiferencialKeystream(; amostras::Integer = 2000, tamanho_bloco::Integer = 32) =
    AtaqueDiferencialKeystream(Int(amostras), Int(tamanho_bloco))

nome(::AtaqueDiferencialKeystream) = "Diferencial do keystream (delta fixo no IV)"

function _xor_buffers(a::AbstractVector{UInt8}, b::AbstractVector{UInt8})
    return UInt8[a[i] ⊻ b[i] for i in eachindex(a)]
end

function executar(a::AtaqueDiferencialKeystream, alvo::AlvoCriptografico)
    chave = random_bytes(length(chave_de_teste(alvo)))
    tam_iv = tamanho_iv(alvo)

    delta = zeros(UInt8, tam_iv)
    delta[random_int(0, tam_iv-1)+1] = UInt8(1 << random_int(0, 7))

    total_bits = a.tamanho_bloco * 8
    sempre_zero = trues(total_bits)
    sempre_um = trues(total_bits)
    deltas = Set{String}()

    for _ in 1:(a.amostras)
        iv = random_bytes(tam_iv)
        iv2 = _xor_buffers(iv, delta)

        ks1 = gerar_keystream_bruto(alvo, chave, iv, "enc", a.tamanho_bloco)
        ks2 = gerar_keystream_bruto(alvo, chave, iv2, "enc", a.tamanho_bloco)
        d = _xor_buffers(ks1, ks2)
        push!(deltas, bytes2hex(d))

        for p in 0:(length(d)-1)
            byte = d[p+1]
            for b in 0:7
                idx = p * 8 + b
                if ((byte >> b) & 1) == 1
                    sempre_zero[idx+1] = false
                else
                    sempre_um[idx+1] = false
                end
            end
        end
    end

    bits_fixos = 0
    for idx in 1:total_bits
        if sempre_zero[idx] || sempre_um[idx]
            bits_fixos += 1
        end
    end
    colisoes = a.amostras - length(deltas)

    vulneravel = bits_fixos > 0 || colisoes > 0

    return ResultadoAtaque(
        nome(a),
        vulneravel,
        vulneravel ? "alta" : "info",
        "delta de 1 bit no IV: $bits_fixos bit(s) de saída fixo(s), $colisoes diferencial(is) repetido(s) em $(a.amostras) amostras",
        Dict{String,Any}("bits_fixos" => bits_fixos, "colisoes" => colisoes),
    )
end
